cfe_internal/core/watchdog/templates/watchdog.mustache

```diff
@@ -1,8 +1,320 @@
+      etimes = ((a[1]*24+a[2])*60 + $2) * 60 + $3;
+    } else {
+      etimes = ($1*60 + $2) * 60 + $3;
+    }
+  }
+ if (etimes > 300) {
+   print
+}
+}')
+if [ ${#observation} -gt 0 ]; then
+    count=$(expr 1 + $(echo -n "${observation}" | wc -l))
+    echo "$(date) Found ${count} cf-agent processes running longer than 300s" >> ${LOGFILE}
+    echo "- ${count} cf-agent processes running longer than 300s" >> ${COLLECTION_REPORT}
+    PATHOLOGY_COUNT=$((${PATHOLOGY_COUNT}+1))
+fi
+
+# Pathology #4: High number of concurrent cf-agent processes
+# While cf-agent is designed to be able to run concurrently, having many
+# concurrent cf-agent processes may indicate an an issue like an agent pile up
+observation=$(ps -e -o pid,etime,args | awk '/\/var\/cfengine\/bin\/cf-agent/ {print}')
+if [ ${#observation} -gt 0 ]; then
+    count=$(expr 1 + $(echo -n "${observation}" | wc -l))
+    if [ ${count} -gt 3 ]; then
+      echo "$(date) Found ${count} concurrently running agents" >> ${LOGFILE}
+      echo "- ${count} concurrently running cf-agent processes" >> ${COLLECTION_REPORT}
+      PATHOLOGY_COUNT=$((${PATHOLOGY_COUNT}+1))
+    fi
+fi
+
+# Pathology #5: cf-check has encountered a critical issue
+# This indicates that there are one or more integrity issues
+if [ -x /var/cfengine/bin/cf-check ]; then
+    observation=$(/var/cfengine/bin/cf-check diagnose /var/cfengine/state/*.lmdb)
+    if [ $? -ne 0 ]; then
+      echo "$(date) cf-check observed critical integrity issues" >> ${LOGFILE}
+      echo "- cf-check observed critical integrity issues" >> ${COLLECTION_REPORT}
+      echo "  #+begin_example\n${observation}\n#+end_example" >> ${COLLECTION_REPORT}
+    fi
+fi
+
+if [ "${PATHOLOGY_COUNT}" -gt "${PATHOLOGY_THRESHOLD}" ]; then
+
+    echo "$(date) Found ${PATHOLOGY_COUNT} symptoms, threshold (${PATHOLOGY_THRESHOLD}) breached." | tee -a ${LOGFILE}
+    echo "* Observations" >> ${COLLECTION_REPORT}
+
+    ps auxwww > "${COLLECTION_DIR}/ps_auxwww.txt"
+    echo "- [[./ps_auxwww.txt][~ps auxwww~]]" >> ${COLLECTION_REPORT}
+
+    ps -elf > "${COLLECTION_DIR}/ps_-elf.txt"
+    echo "- [[./ps_-elf.txt][~ps -elf~]]" >> ${COLLECTION_REPORT}
+
+    find "${CFENGINE_WORKDIR}" >> "${COLLECTION_DIR}/find__var_cfengine.txt"
+    echo "- [[./find__var_cfengine.txt][=/var/cfengine= file list]] - Before remediation" >> ${COLLECTION_REPORT}
+
+    tar -c -f "${COLLECTION_DIR}/sys.workdir-before-remediation.tar" -C "${CFENGINE_WORKDIR}" state outputs
+    echo "- [[./sys.workdir-before-remediation.tar][=sys.workdir-before-remediation.tar=]] - CFEngine WORKDIR artifacts before remediation" >> ${COLLECTION_REPORT}
+
+    ps -efl | grep cf- > "${COLLECTION_DIR}/cf-procs.txt"
+    echo "- [[./cf-procs.txt][~ps -efl | grep cf-~]]" >> ${COLLECTION_REPORT}
+
+    echo "$(date) Initiating apoptosis" | tee -a ${LOGFILE}
+    while IFS= read -r proc
+    do
+        _PID=$(echo $proc | awk '{print $4}')
+        _PROC=$(echo $proc | awk '{print $15}')
```

Comment:
How portable are these field indices?

Reply:
I don't know. But this watchdog is only used on aix. it's not really clear by the paths or anything, but the watchdog for linux (the cron.d implementation) uses templates/cfengine_watchdog.mustache. I have not tried to tidy that stuff in this PR.